data/transactions/logic/resources.go

```diff
@@ -0,0 +1,359 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// resources contains a catalog of available resources. It's used to track the
+// apps, assets, and boxes that are available to a transaction, outside the
+// direct foreign array mechanism.
+type resources struct {
+	// These resources were created previously in the group, so they can be used
+	// by later transactions.
+	createdAsas []basics.AssetIndex
+	createdApps []basics.AppIndex
+
+	// These resources have been used by some txn in the group, so they are
+	// available. These maps track the availability of the basic objects (often
+	// called "params"), not the "cross-product" objects (which are tracked
+	// below)
+	sharedAccounts map[basics.Address]struct{}
+	sharedAsas     map[basics.AssetIndex]struct{}
+	sharedApps     map[basics.AppIndex]struct{}
+	// We need to carefully track the "cross-product" availability, because if
+	// tx0 mentions an account A, and tx1 mentions an ASA X, that does _not_
+	// make the holding AX available
+	sharedHoldings map[ledgercore.AccountAsset]struct{}
```

Comment:
`TestAppSharing` and `TestAssetSharing` sort of do it. In each, the test after the explanation:
```
	// In contrast, here there's no help from v9, because the second tx is
	// reading a holding for a different account.
```
tests it, because the app or asset is made available in tx0, but the account is from tx1.

On the other hand, I could also add tests such that the app/asset is in tx0, and account is in tx1, but it's tx2 that tries to access the cross-product.

Reply:
thanks, lets add this cross-product test

Reply:
This commit has adds those tests (but will change after I rebase later today!)

https://github.com/algorand/go-algorand/pull/5035/commits/695a2a3b6184cde326cc627ad6bc94a10467c24c